core/treepath/treepath.go

```diff
@@ -0,0 +1,173 @@
+// * Parse("1.2:1.0.2:3") => [1,2,0,2,2,2]
+func Parse(path string) (Treepath, error) {
+	out := Treepath{}
+	curState := variationState
+	buf := &strings.Builder{}
+	prevChar := 0
+
+	convertBuffer := func(idx int) (int, error) {
+		if buf.Len() == 0 {
+			return 0, fmt.Errorf("error parsing fragment %q at index %d: separator '.' must be proceeded by digit", path, idx)
+		}
+		n, err := strconv.Atoi(buf.String())
+		if err != nil {
+			return 0, fmt.Errorf("error parsing number in fragment %q at index %d: %v", path, idx, err)
+		}
+		buf = &strings.Builder{}
+		return n, nil
+	}
+
+	// Use a sentinal newline character so that it's clear when to flush the final
+	// charactrs.
+	path += "\n"
+
+	//
+	// The rough approach is to have a simple parser modeled as a 2-state DFA.
+	//
+	// VARIATION: means we're parsing a normal variation number. We flush the
+	// buffer data when we hit a '.' rune.
+	//
+	// REPEAT: means we have seen a ':' instead of a '.', so we're going to repeat
+	// the previous number once we parse the next number.
+	//
+	//  --'.'--- <------------
+	//  |      |             |
+	//  V      ^             ^
+	// VARIATION --':'-->  REPEAT
+	//
+	for idx, c := range path {
+		if unicode.IsDigit(c) {
+			buf.WriteRune(c)
+		} else if c == '.' || c == '\n' {
+			n, err := convertBuffer(idx)
+			if err != nil {
+				return nil, err
+			}
+
+			// State transitions
+			if curState == variationState {
+				// Variation => Variation (& flush)
+				out = append(out, n)
+			} else if curState == repeatState {
+				// Repeat => Variation (& flush)
+				for i := 0; i < n; i++ {
+					out = append(out, prevChar)
+				}
+				curState = variationState
+				prevChar = 0
+			} else {
+				return nil, fmt.Errorf("error parsing fragment %q at index %d: unknown state %v", path, idx, curState)
+			}
+		} else if c == ':' {
+			n, err := convertBuffer(idx)
+			if err != nil {
+				return nil, err
+			}
+
+			// State transitions
+			if curState == variationState {
+				// Variation => Repeat
+				prevChar = n
+				curState = repeatState
+			} else if curState == repeatState {
+				// Repeate => Repeat: Invalid
+				return nil, fmt.Errorf("error parsing fragment %q at index %d: repeat char ':' cannot be followed by another repeat ':'", path, idx)
+			} else {
+				return nil, fmt.Errorf("error parsing fragment %q at index %d: unknown state %v", path, idx, curState)
+			}
+		} else {
+			return nil, fmt.Errorf("error parsing fragment %q at index %d: unexpected character %q", path, idx, c)
+		}
+	}
+	return out, nil
+}
+
+// Apply applies a treepath to a node, moving down a tree of moves. This
+// takes the variation listed in the treepath until either:
+//
+// 1. There are no more variation numbers in the treepath.
+// 2. There is not a child with the given variation number.
+func (tp Treepath) Apply(n *game.Node) *game.Node {
+	curNode := n
+	for _, v := range tp {
+		if v < len(curNode.Children) {
+			fmt.Printf("var %v\n", v)
```

Comment:
Is this output intentional, basically info logging going to stdout for now?

Reply:
Done